src/components/SearchInput/SearchInput.tsx

```diff
@@ -0,0 +1,130 @@
+import {
+  ChangeEvent,
+  ComponentProps,
+  FocusEvent,
+  MouseEvent,
+  ReactNode,
+  useEffect,
+  useRef,
+  useState,
+} from "react"
+import { useKeyPress } from "../../hooks/useKeyPress"
+import styles from "./SearchInput.module.css"
+
+type Omitted = "onChange" | "type"
+
+interface SearchInputProps extends Omit<ComponentProps<"input">, Omitted> {
+  "aria-label": string
+  searchIcon?: ReactNode
+  clearIcon?: ReactNode
+  debounceDelay?: number
+  value?: string
+  onSearch?: (searchTerm: string) => void
+  showSearchButton?: boolean
+}
+
+const shouldInvokeDebounce = (
+  searchTerm: string,
+  showSearchButton?: boolean,
+) => {
+  return Boolean(searchTerm && !showSearchButton)
+}
+
+export function SearchInput({
+  searchIcon = <span>🔍</span>,
+  clearIcon = <span>❎</span>,
+  onSearch,
+  debounceDelay = 300,
+  showSearchButton,
+  value = "",
+  ...inputProps
+}: SearchInputProps) {
+  const [searchTerm, setSearchTerm] = useState("")
+  const [clearButtonHidden, setClearButtonHidden] = useState(true)
+  const timer = useRef<NodeJS.Timeout>()
+
+  useKeyPress(["Enter"], () => {
+    onSearch?.(searchTerm)
+  })
+
+  useEffect(() => {
+    setSearchTerm(value)
+  }, [value])
+
+  useEffect(() => {
+    if (shouldInvokeDebounce(searchTerm, showSearchButton)) {
+      timer.current = setTimeout(() => onSearch?.(searchTerm), debounceDelay)
+    }
+    return () => {
+      clearTimeout(timer.current)
+    }
+  }, [searchTerm, onSearch, showSearchButton, debounceDelay])
```

Comment:
I don't think this implementation of `debounce` translates accurately to the original one in that it changes its behavior from imperative to reactive. `debounce` is associated with an input action that _must_ take place before the outgoing action is triggered, but here, changing `onSearch`, `showSearchButton` or `debounceDelay` could trigger `onSearch` to be called, but the input action part of it is exclusively related to the `searchTerm`. 
So, `searchTerm` should be the only thing that could possibly trigger `onSearch`.

In practice it's unlikely that these other variables would be changed on the fly, so, more philosophy here.